### PR TITLE
Improved support for frontmatter

### PR DESCRIPTION
**Pull request by ItsMeBrianD**
### Description

This PR addresses #294; and improves the support for Markdown Frontmatter.

MDSvex already provides any frontmatter to the component (in this case, `+page.md`), so this PR primarily addresses adding things like a page title, description, and opengraph information.

Because this touches preprocess, I took the opportunity to split it up into several files, functionality should not be impacted, but it makes things a little easier to reason with when looking at a specific part of the pipeline.

### Checklist
- [ ] ~For UI or styling changes, I have added a screenshot or gif showing before & after~
- [x] I have added a [changeset](https://github.com/evidence-dev/evidence/blob/main/CONTRIBUTING.md#adding-a-changeset)

### CODE REVIEW

**Comment by ItsMeBrianD** [comment]
Fixed the 2 listed issues;

With regard to #85, I think there will need to be a little more clarity on the goal and use-cases. In theory it could be handled either at build-time or at run-time, as the logic to extract the frontmatter will be basically the same (e.g. regex match on `---` and then parse the yaml content).

File: packages/preprocess/src/process-queries.cjs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+         }
+    });
+    return queryIds;
+}
+
+/**
+ * @type {(componentDevelopmentMode: boolean) => import("svelte-preprocess/dist/types").PreprocessorGroup}
+ */
+const processQueries = (componentDevelopmentMode) => {
+    let queryIdsByFile = {}
```
Comment (ud3sh) [inline packages/preprocess/src/process-queries.cjs:144]: This `queryIdsByFile` was bit of a strange "workaround" as we needed remember state between different pre-process phases (without having to load things twice from what I recall).   I think this works neatly with way you refactored this, under the assumption that the order of the array returned by `evidencePreprocess` doesn't matter.   I think it doesn't but I am not sure.

File: sites/example-project/src/pages/frontmatter/+page.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+---
```
Comment (ud3sh) [inline sites/example-project/src/pages/frontmatter/+page.md:1]: +1 for example!

**Review by ud3sh — APPROVED** [review]

File: sites/example-project/src/pages/frontmatter/+page.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+| `title`          | changes the name of the tab, and also adds a header to your page                                                             |
+| `description`    | is used for search engines                                                                                                   |
+| `og`             | changes how your link shows up when shared on things like Slack, Facebook, Twitter, Discord, etc                             |
+| `og.title`       | changes the title that appears in the embed; if this is not specified, but `title` is, then `title` is used (and vise versa) |
+| `og.description` | changes the body of the embed                                                                                                |
+| `og.image`       | will appear in the embed if specified, but it is not required.                                                               |
+
+
+You can access any of your frontmatter values by surrounding them with `{}`.
+For example, this page has `some_arbitrary_frontmatter` included in it's own frontmatter, which has a value of `{some_arbitrary_frontmatter}`
```
Comment (mcrascal) [inline sites/example-project/src/pages/frontmatter/+page.md:37]: Can we update this example to show this value being printed into the page?

Comment (ItsMeBrianD) [reply sites/example-project/src/pages/frontmatter/+page.md:37]: This should render as 
> For example, this page has `some_arbitrary_frontmatter` included in it's own frontmatter, which has a value of `This is arbitrary, and can be anything`

Comment (ItsMeBrianD) [reply sites/example-project/src/pages/frontmatter/+page.md]: This is fixed on the latest deployment

**Comment by mcrascal** [comment]
This looks really good. 

1. We'll need docs: I'd add this to the markdown reference page, and I'd include a sentence or two on the `core-concepts/pages` with a link over to the reference. 
2. I really like the open graph support. I think in a future iteration of that it'd be great to default back to an image of the first chart on the page. This is beyond the scope of this frontmatter issue though, so I'll open something separate for that.

**Review by mcrascal — APPROVED** [review]

**Comment by archiewood** [comment]
Hey @ItsMeBrianD, while I've been testing the standalone SQL files PR, I've run into an issue (or at least a thing that worked unexpectedly)

A pretty common pattern in templated pages is to add a title named after the parameter.

ie if you have 
```
|-- index.md
`-- [categories] /
       `-- index.md
```
Then you title it like this

```markdown
# {$page.params.category}

Some other content
```

I assumed this would work in the frontmatter like this
```
---
title: {$page.params.category}
---

Some other content
```

which actually gives: 
image

I assume the js object is being converted into a string before it reaches the page somehow?

**Comment by archiewood** [comment]
This also brings up a more general question about js support in frontmatter. (We do not have to sovle this now, just thoughts)

It would be great if we can eliminate as much of the need for people to add `<script>` tags as possible.

**Should we support arbitrary js in frontmatter?**

I quite often end up getting an associated field from a table like this.
```
<script>
  let company_id = companies.filter(d=>d.company===$page.params.company)[0].company_id
<script>
```
```
---
company_id: companies
- filter: company
- equals: {$page.params.company}
- column: company_id
- row: 0
---
```

or just

```
---
company_id:  {companies.filter(d=>d.company===$page.params.company)[0].company_id}
---
```

This could be a bad idea for reasons I have not thought of!

**Comment by ItsMeBrianD** [comment]
Regarding the title; that might work with this syntax:
```
---
title: $page.params.category
---
```
If it doesn't then this probably isn't supported right now (but could be), worth exploring for sure.

Adding variables / filters in the frontmatter could be nice; but might get a little wonky - we'll have to explore. This probably deserves it's own issue to make sure that we can have the discussion and nail it down before writing the code.

**Pull request merged**